core/src/main/java/org/apache/iceberg/DeleteFileIndex.java

```diff
@@ -117,13 +117,18 @@ public Iterable<DeleteFile> referencedDeleteFiles() {
     return deleteFiles;
   }
 
-  private StructLikeWrapper newWrapper(int specId) {
-    return StructLikeWrapper.forType(partitionTypeById.get(specId));
+  private Map<Integer, ThreadLocal<StructLikeWrapper>> wrappers(Map<Integer, PartitionSpec> specs) {
+    Map<Integer, ThreadLocal<StructLikeWrapper>> wrappers = Maps.newHashMap();
+    specs.forEach((specId, spec) -> wrappers.put(specId, newWrapper(specId)));
+    return wrappers;
+  }
+
+  private ThreadLocal<StructLikeWrapper> newWrapper(int specId) {
+    return ThreadLocal.withInitial(() -> StructLikeWrapper.forType(partitionTypeById.get(specId)));
   }
 
   private Pair<Integer, StructLikeWrapper> partition(int specId, StructLike struct) {
-    ThreadLocal<StructLikeWrapper> wrapper =
-        wrapperById.computeIfAbsent(specId, id -> ThreadLocal.withInitial(() -> newWrapper(id)));
```

Comment:
The cost of `computeIfAbsent` in `ConcurrentHashMap` may be non-trivial and this code is invoked for each data file in tight loops, potentially from multiple threads. I noticed this while profiling job planning. There are not that many specs in the table and creating a wrapper per spec is nothing compared to using a thread-safe collection and accessing it for each data file that must be planned.

Reply:
Yeah, we discussed that issue in a previous PR. I doubt the performance difference would be visible now given that we read manifest entries locally (there is non-trivial cost for decoding and other things that dominates). I noticed it when profiling distributed planning where all the data files are already loaded in memory. It is not critical, but it appeared on a few flamegraphs. I just don't really see value in `ConcurrentHashMap` in this use case.